Summary details - do not merge

BenjiKan opened the pull request:
These features partially resolve #199. This branch adds several useful summary statistics and two plots to the qa-summary page, including a fractional residuals vs. flux plot and median calibrated flux vs. fiber flux plot (all sourced from @deisenstein's notebook at `/project/projectdirs/desi/users/eisenste/NightWatchVisualizationsDev.ipynb` ). An example can be seen at https://data.desi.lbl.gov/desi/users/benjikan/nightwatch/20210206/00074938/qa-summary-00074938.html, and now also includes the following summary text:
```
570.2 second SCIENCE (sv1elgqso)
Tile 80673 at RA, Dec 85.5, -20.2, Galactic l,b 224.3, -24.0
Airmass 1.66, Hour Angle -10 deg at UTC 03:21:43
Moon is set
r=20 stars yielding 137.0 counts/ks in R with 0.084 fractional residual
Sky yielding 109.9 counts/ks in R
```
(note that if the moon's elevation angle were >-6 deg, the "Moon is set" line would instead read "Moon is 12% full, 123 deg away, 12 deg above the horizon")

The plotting and calculations are packaged in new functions in plots/camfiber.py, while new fields are added in webpages/summary.py and webpages/templates/summary.html. 

The path to spectro files and the nightwatch directory are currently hardcoded (in [lines 356-357](https://github.com/desihub/nightwatch/compare/summary-details?expand=1#diff-f5f13dca6233e5fbe16353bba9501f15f90cf6696e19a3deb4e5dedc1fbcc308R356)) into camfiber.py, since I couldn't find an existing function or global path to either. A simple fix would be to pass the NW directory path as an explicit argument into the function, but I don't know if there's a preferred solution. I can add more documentation to these new functions when that is figured out.

deisenstein commented:
Looking nice, thanks!

Small tweaks:

No commas between the two numbers for RA and Dec and between l and b, i.e.,
85.5 -20.2.  And let's change to "-20.2 and Galactic" (no comma).

Can you change the aspect ratio of the focal plane plot to make it a
circle?  That will make it less tall.  And probably make it the same width
as the plots above.

And then change the aspect ratio of the flux vs flux plot so that when
these two plots are side-by-side, the width is the same as the top row and
the spectrum plot.  At present, it's a fair bit wider.

A bug in my code that will have propagated to yours: I look for the
standard star target bits in SV1_DESI_TARGET, but that column name will
disappear soon.  We need to get the right mechanism for this.

There may be a need to make these plots more robust, i.e., if my code
doesn't work, does the plot just quietly not exist?  Or does it crash the
whole summary page?

And to confirm: these plots don't exist for the calibration lamps?

Thanks,
Daniel


On Mon, Mar 8, 2021 at 5:36 PM BenjiKan <notifications@github.com> wrote:

> These features partially resolve #199
> <https://github.com/desihub/nightwatch/issues/199>. This branch adds
> several useful summary statistics and two plots to the qa-summary page,
> including a fractional residuals vs. flux plot and median calibrated flux
> vs. fiber flux plot (all sourced from
> /project/projectdirs/desi/users/eisenste/NightWatchVisualizationsDev.ipynb
> ). An example can be seen at
> https://data.desi.lbl.gov/desi/users/benjikan/nightwatch/20210206/00074938/qa-summary-00074938.html,
> and now also includes the following summary text:
>
> 570.2 second SCIENCE (sv1elgqso)
> Tile 80673 at RA, Dec 85.5, -20.2, Galactic l,b 224.3, -24.0
> Airmass 1.66, Hour Angle -10 deg at UTC 03:21:43
> Moon is set
> r=20 stars yielding 137.0 counts/ks in R with 0.084 fractional residual
> Sky yielding 109.9 counts/ks in R
>
> (note that if the moon's elevation angle were >-6 deg, the "Moon is set"
> line would instead read "Moon is 12% full, 123 deg away, 12 deg above the
> horizon")
>
> The plotting and calculations are packaged in new functions in
> plots/camfiber.py, while new fields are added in webpages/summary.py and
> webpages/templates/summary.html.
>
> The path to spectro files and the nightwatch directory are currently
> hardcoded (in lines 356-357
> <https://github.com/desihub/nightwatch/compare/summary-details?expand=1#diff-f5f13dca6233e5fbe16353bba9501f15f90cf6696e19a3deb4e5dedc1fbcc308R356>)
> into camfiber.py, since I couldn't find an existing function or global path
> to either. A simple fix would be to pass the NW directory path as an
> explicit argument into the function, but I don't know if there's a
> preferred solution. I can add more documentation to these new functions
> when that is figured out.
> ------------------------------
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/desihub/nightwatch/pull/220
> Commit Summary
>
>    - add median flux plot, detailed condition text
>    - update summarytext for clarity, reorder plots
>    - html formatting
>    - formatting
>    - fixed typo
>
> File Changes
>
>    - *M* py/nightwatch/plots/camfiber.py
>    <https://github.com/desihub/nightwatch/pull/220/files#diff-f5f13dca6233e5fbe16353bba9501f15f90cf6696e19a3deb4e5dedc1fbcc308>
>    (158)
>    - *M* py/nightwatch/webpages/summary.py
>    <https://github.com/desihub/nightwatch/pull/220/files#diff-2c9a8896838ea2b1be808facd68f7dc4efb85cfb7463352340f7172fa009c47f>
>    (13)
>    - *M* py/nightwatch/webpages/templates/summary.html
>    <https://github.com/desihub/nightwatch/pull/220/files#diff-7a34f1cc185d7618135b58fc76aa1ebb2898526331599c3b938d26ca8b458285>
>    (19)
>
> Patch Links:
>
>    - https://github.com/desihub/nightwatch/pull/220.patch
>    - https://github.com/desihub/nightwatch/pull/220.diff
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/desihub/nightwatch/pull/220>, or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAPQRGWRBEBBMLSOA5IRO4TTCVGO5ANCNFSM4Y2IPFJA>
> .
>

BenjiKan commented:
Thanks for the comments, I've made the first three changes you mentioned and updated https://data.desi.lbl.gov/desi/users/benjikan/nightwatch/20210206/00074938/qa-summary-00074938.html. The widths of the first four plots don't line up perfectly with the spectra plot (due to the toolbar), but they are much closer now.

How can we find the standard star target bits, if not through SV1_DESI_TARGET? As for robustness, I've added and tested a try-except block in case something in the calculations break.

I'm not very familiar with the calibration lamps, so I'm not quite sure what you're asking. I see you have a separate notebook for them in your directory, is that what you're referencing? If so, those plots are not currently included on this summary page.

deisenstein commented:
Thanks.

For the standard stars, the problem is that with today's tiles, the field
is called SV1_DESI_TARGET, but future tiles will use different names, e.g.,
SV2_DESI_TARGET, etc.  I'm 99% sure that there's a function in the desi
code base that will tell us where to find the standard star bits for any
tile.  Stephen, can you let us know how to do that?

For the calibration lamps, I just mean one of the earlier exposures on the
night -- they're the ones called obstype ZERO, ARC, FLAT.  These plots
would be nonsense for those, so we just want to skip them.  That may
already be happening.

A couple more tweaks: let's use the titles "Fractional residuals of
standard star R counts vs flux" and "Median Calibrated Flux vs Fiberflux
for Standard Stars".

And on my screen, the focal plane is still a little ellipsoidal (too tall),
so I think you can trim both plots to be a little shorter.

Stephen, I think this is ready for at least a quick conceptual review.

Daniel

On Mon, Mar 8, 2021 at 9:38 PM BenjiKan <notifications@github.com> wrote:

> Thanks for the comments, I've made the first three changes you mentioned
> and updated
> https://data.desi.lbl.gov/desi/users/benjikan/nightwatch/20210206/00074938/qa-summary-00074938.html.
> The widths of the first four plots don't line up perfectly with the spectra
> plot (due to the toolbar), but they are much closer now.
>
> How can we find the standard star target bits, if not through
> SV1_DESI_TARGET? As for robustness, I've added and tested a try-except
> block in case something in the calculations break.
>
> I'm not very familiar with the calibration lamps, so I'm not quite sure
> what you're asking. I see you have a separate notebook for them in your
> directory, is that what you're referencing? If so, those plots are not
> currently included on this summary page.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/desihub/nightwatch/pull/220#issuecomment-793288680>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAPQRGTIJB37TRJJ33HBJW3TCWC2RANCNFSM4Y2IPFJA>
> .
>

BenjiKan commented:
Ah yes, these calculations only run if it is a SCIENCE exposure. NERSC is acting up for me again, but I've just tweaked the code for titles and plot sizing, which should propagate once I'm able to connect again.

deisenstein commented:
Perfect, thanks.

On Mon, Mar 8, 2021 at 11:29 PM BenjiKan <notifications@github.com> wrote:

> Ah yes, these calculations only run if it is a SCIENCE exposure. NERSC is
> acting up for me again, but I've just tweaked the code for titles and plot
> sizing, which should propagate once I'm able to connect again.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/desihub/nightwatch/pull/220#issuecomment-793370843>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAPQRGWWP4GG4X4EX2BJYWDTCWP33ANCNFSM4Y2IPFJA>
> .
>

sbailey commented on py/nightwatch/plots/camfiber.py:
```diff
@@ -291,3 +296,156 @@ def plot_per_fibernum(cds, attribute, cameras, titles={},
+    fa.sort(order="FIBER")
+    print(file, len(fa))
+    try:
+        standard = goodfiber&(fa["OBJTYPE"]=='TGT')&((fa['SV1_DESI_TARGET']&0xe00000000)>0)    # Bits 33..35
```
`desispec.fluxcalibration.isStdStar(fa)` is the future-proof version of this, i.e. it is supposed to handle SV2_DESI_TARGET, SV3_DESI_TARGET, DESI_TARGET, etc. (and if that breaks, we'll fix it there so that code light nightwatch doesn't also have to change).

sbailey commented on py/nightwatch/plots/camfiber.py:
```diff
@@ -291,3 +296,156 @@ def plot_per_fibernum(cds, attribute, cameras, titles={},
+    phase = (1.0-np.arccos(np.dot(moon,sun))/np.pi)   # 0 = moon, sun together; 1 = opposed.
+    return np.arcsin(np.dot(zenith,sun))*180.0/np.pi, phase
+
+def plot_fractionalresidual(fiber, header, expos, position=False, plot_height=500, plot_width=500):
```
Please add docstrings to all functions defining their inputs and outputs so that someone would know how to call the function without having to read the code.  e.g. what is `fiber`, and `expos`, and what would `position=True` do?

sbailey reviewed (commented):
Thanks for adding this @BenjiKan and for seed code from @deisenstein . These will very useful plots to have featured on the summary page.  I put some small comment inline, but the main code organization change request is to separate the calculations from the plotting/display of those results.  i.e. much of this code should be triggered by `nightwatch.run.run_qa -> nightwatch.qa.runner.QARunner.run()`, writing the outputs to a file, which is then read back in and passed to the plotting code for display.  That separation of calculation vs. display solves two problems:
  * avoiding the hardcoded directories you added here (and correctly flagged as needing attention), because when the QA is being run, that code does know where the input and output directories are
  * enabling future code to do things like slurp in a bunch of nightwatch files and make a plot of spectroscopic sky counts or R=20 star counts vs. time.  Right now those numbers get calculated but then buried in an HTML file where a human can see them but it's not easily accessed by other code.

That being said, we got a bit out of hand with "elegant" object oriented QAblah design that backed us into a corner that this calculation may not fit so well.  I think this calculation boils down to collecting / calculating a bunch of scalars per exposure (moon altitude, star counts/ks, sky counts/ks, ...) plus the list of target indices to be used for plotting.  Is that correct?  While I think a little bit more about how to fit this in, can you restructure this code into
  1. calculation: a function that takes a list of inputs that it needs and returns a dictionary of results
  2. plotting: a function that takes `qadata['PER_CAMFIBER']` and that previous dictionary of results to make the plots without having to do any additional I/O or fitting?

Depending upon how that return dictionary from the calculation looks, this might fit into a new "PER_EXP" QA (it is calculating a bunch of numbers per exposure, which is what the not-yet-implemented PER_EXP QA was intended for...).  Or maybe it becomes a different file or a special-case HDU of the qa-EXPID.fits file (unfortunately FITS is not good at storing json-like nested-dictionaries+lists structures which would be ideal for this, but it would also be nice to keep all the results together in a single file...)

BenjiKan commented:
Thanks, I've separated into the calculation and plotting functions, leaving both are in plots/camfiber.py for now. Once we move the calculation into QA, I can simplify the parameters for the plotting function.

The calculation does produce several scalars for the summary text, but it also produces multiple arrays for the plots—since some of these arrays rely on the coordinates file and not just `qadata['PER_CAMFIBER']`, the calculation function returns a separate dict for these arrays than for the scalars. I'm not sure if the coordinates file would be considered part of the I/O.

On a separate note: now that we have this dictionary of scalar values, it might be preferable to construct the summary text on the HTML template itself; would that be better than doing so in the plotting function? That would make any changes and additions to this text slightly more involved, but might improve readability of the HTML code.
